File: library-extensions/src/main/java/com/mikepenz/fastadapter_extensions/swipe/SimpleSwipeDragCallback.java
package com.mikepenz.fastadapter_extensions.swipe;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import com.mikepenz.fastadapter_extensions.drag.ItemTouchCallback;
import com.mikepenz.fastadapter_extensions.drag.SimpleDragCallback;

/**
 * Created by Mattias on 2016-02-13.
 */
public class SimpleSwipeDragCallback extends SimpleDragCallback {

    private final SimpleSwipeCallback simpleSwipeCallback;

    public SimpleSwipeDragCallback(ItemTouchCallback itemTouchCallback, SimpleSwipeCallback.ItemSwipeCallback itemSwipeCallback, Drawable leaveBehindDrawable) {
        this(itemTouchCallback, itemSwipeCallback, leaveBehindDrawable, ItemTouchHelper.LEFT);
    }

    public SimpleSwipeDragCallback(ItemTouchCallback itemTouchCallback, SimpleSwipeCallback.ItemSwipeCallback itemSwipeCallback, Drawable leaveBehindDrawable, int swipeDirs) {
        this(itemTouchCallback, itemSwipeCallback, leaveBehindDrawable, swipeDirs, Color.RED);
    }

    public SimpleSwipeDragCallback(ItemTouchCallback itemTouchCallback, SimpleSwipeCallback.ItemSwipeCallback itemSwipeCallback, Drawable leaveBehindDrawable, int swipeDirs, @ColorInt int bgColor) {
        super(itemTouchCallback);
        setDefaultSwipeDirs(swipeDirs);
        simpleSwipeCallback = new SimpleSwipeCallback(itemSwipeCallback, leaveBehindDrawable, swipeDirs, bgColor);
    }

    public SimpleSwipeDragCallback withLeaveBehindSwipeLeft(Drawable d) {
        setDefaultSwipeDirs(super.getSwipeDirs(null, null) | ItemTouchHelper.LEFT);
        simpleSwipeCallback.withLeaveBehindSwipeLeft(d);
        return this;
    }

    public SimpleSwipeDragCallback withLeaveBehindSwipeRight(Drawable d) {
        setDefaultSwipeDirs(super.getSwipeDirs(null, null) | ItemTouchHelper.RIGHT);
        simpleSwipeCallback.withLeaveBehindSwipeRight(d);
        return this;
    }

    public SimpleSwipeDragCallback withHorizontalMarginDp(Context ctx, int dp) {
        simpleSwipeCallback.withHorizontalMarginDp(ctx, dp);
        return this;
    }

    public SimpleSwipeDragCallback withHorizontalMarginPx(int px) {
        simpleSwipeCallback.withHorizontalMarginPx(px);
        return this;
    }

    public SimpleSwipeDragCallback withBackgroundSwipeLeft(@ColorInt int bgColor) {
        simpleSwipeCallback.withBackgroundSwipeLeft(bgColor);
        return this;
    }

    public SimpleSwipeDragCallback withBackgroundSwipeRight(@ColorInt int bgColor) {
        simpleSwipeCallback.withBackgroundSwipeRight(bgColor);
        return this;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        simpleSwipeCallback.onSwiped(viewHolder, direction);
    }

    @Override
    public int getSwipeDirs(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        return simpleSwipeCallback.getSwipeDirs(recyclerView, viewHolder);
    }


    @Override
    public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
        simpleSwipeCallback.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        //Happen to know that our direct parent class doesn't (currently) draw anything...
        //super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
    }
}

File: library-extensions-utils/src/main/java/com/mikepenz/fastadapter/swipe_drag/SimpleSwipeDragCallback.kt
@file:Suppress("PackageName", "PackageNaming")

package com.mikepenz.fastadapter.swipe_drag

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.Drawable
import androidx.annotation.ColorInt
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.drag.ItemTouchCallback
import com.mikepenz.fastadapter.drag.SimpleDragCallback
import com.mikepenz.fastadapter.swipe.SimpleSwipeCallback

/**
 * Created by Mattias on 2016-02-13.
 */
class SimpleSwipeDragCallback @JvmOverloads constructor(
        itemTouchCallback: ItemTouchCallback,
        itemSwipeCallback: SimpleSwipeCallback.ItemSwipeCallback,
        leaveBehindDrawable: Drawable?,
        swipeDirs: Int = ItemTouchHelper.LEFT,
        @ColorInt bgColor: Int = Color.RED
) : SimpleDragCallback(itemTouchCallback) {

    private val simpleSwipeCallback: SimpleSwipeCallback
    private var defaultSwipeDirs: Int = 0

    init {
        setDefaultSwipeDirs(swipeDirs)
        simpleSwipeCallback = SimpleSwipeCallback(itemSwipeCallback, leaveBehindDrawable, swipeDirs, bgColor)
    }

    override fun setDefaultSwipeDirs(defaultSwipeDirs: Int) {
        this.defaultSwipeDirs = defaultSwipeDirs
        super.setDefaultSwipeDirs(defaultSwipeDirs)
    }

    fun withLeaveBehindSwipeLeft(d: Drawable): SimpleSwipeDragCallback {
        setDefaultSwipeDirs(defaultSwipeDirs or ItemTouchHelper.LEFT)
        simpleSwipeCallback.withLeaveBehindSwipeLeft(d)
        return this
    }

    fun withLeaveBehindSwipeRight(d: Drawable): SimpleSwipeDragCallback {
        setDefaultSwipeDirs(defaultSwipeDirs or ItemTouchHelper.RIGHT)
        simpleSwipeCallback.withLeaveBehindSwipeRight(d)
        return this
    }

    fun withHorizontalMarginDp(ctx: Context, dp: Int): SimpleSwipeDragCallback {
        simpleSwipeCallback.withHorizontalMarginDp(ctx, dp)
        return this
    }

    fun withHorizontalMarginPx(px: Int): SimpleSwipeDragCallback {
        simpleSwipeCallback.withHorizontalMarginPx(px)
        return this
    }

    fun withBackgroundSwipeLeft(@ColorInt bgColor: Int): SimpleSwipeDragCallback {
        setDefaultSwipeDirs(defaultSwipeDirs or ItemTouchHelper.LEFT)
        simpleSwipeCallback.withBackgroundSwipeLeft(bgColor)
        return this
    }

    fun withBackgroundSwipeRight(@ColorInt bgColor: Int): SimpleSwipeDragCallback {
        setDefaultSwipeDirs(defaultSwipeDirs or ItemTouchHelper.RIGHT)
        simpleSwipeCallback.withBackgroundSwipeRight(bgColor)
        return this
    }

    fun withNotifyAllDrops(notifyAllDrops: Boolean): SimpleSwipeDragCallback {
        this.notifyAllDrops = notifyAllDrops
        return this
    }

    fun withSensitivity(f: Float): SimpleSwipeDragCallback {
        simpleSwipeCallback.withSensitivity(f)
        return this
    }

    fun withSurfaceThreshold(f: Float): SimpleSwipeDragCallback {
        simpleSwipeCallback.withSurfaceThreshold(f)
        return this
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        simpleSwipeCallback.onSwiped(viewHolder, direction)
    }

    override fun getSwipeDirs(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
        return simpleSwipeCallback.getSwipeDirs(recyclerView, viewHolder)
    }

    override fun getSwipeEscapeVelocity(defaultValue: Float): Float {
        return simpleSwipeCallback.getSwipeEscapeVelocity(defaultValue)
    }

    override fun getSwipeThreshold(viewHolder: RecyclerView.ViewHolder): Float {
        return simpleSwipeCallback.getSwipeThreshold(viewHolder)
    }

    override fun onChildDraw(
            c: Canvas,
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder,
            dX: Float,
            dY: Float,
            actionState: Int,
            isCurrentlyActive: Boolean
    ) {
        simpleSwipeCallback.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        //Happen to know that our direct parent class doesn't (currently) draw anything...
        //super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
    }
}
